src/viam/sdk/components/board/board.hpp

```diff
@@ -0,0 +1,162 @@
+/// @file components/board/board.hpp
+///
+/// @brief Defines a `Board` component.
+#pragma once
+
+#include <string>
+
+#include <viam/api/component/board/v1/board.pb.h>
+
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/resource/resource_manager.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Board Classes related to the `Board` component.
+
+/// @class BoardSubtype
+/// @brief Defines a `ResourceSubtype` for the `Board` component.
+/// @ingroup Board
+class BoardSubtype : public ResourceSubtype {
+   public:
+    explicit BoardSubtype(const google::protobuf::ServiceDescriptor* service_descriptor);
+    std::shared_ptr<ResourceServer> create_resource_server(
+        std::shared_ptr<ResourceManager> manager) override;
+    std::shared_ptr<Resource> create_rpc_client(std::string name,
+                                                std::shared_ptr<grpc::Channel> chan) override;
+};
+
+/// @class Board board.hpp "components/board/board.hpp"
+/// @brief Represents a physical board with gpio pins, digital interrupts, and analog voltage
+/// reading
+/// @ingroup Board
+///
+/// This acts as an abstract base class to be inherited from by any drivers representing
+/// specific board implementations. This class cannot be used on its own.
+class Board : public Component {
+   public:
+    /// @brief Represents the value received by the registered analog to digital converter (ADC).
+    /// The range and conversion mechanism to voltage will vary depending on the specific ADC
+    /// registered to the pin. Consult your ADC's documentation and Viam's `Board` documentation for
+    /// more details.
+    typedef int32_t analog_value;
+
+    /// @brief Depending on the type of digital interrupt, this can have different meanings. If a
+    /// `basic` (default) interrupt is configured, then this is the total interrupt count. If a
+    /// `servo` interrupt is configured this tracks the pulse width value. Consult Viam's `Board`
+    /// documentation for more details.
+    typedef int64_t digital_value;
+
+    /// @struct board_reader_status
+    /// @brief This contains all of the values for all of the registered analog readers and digital
+    /// interrupts that have been registered on the board.
+    struct status {
+        std::map<std::string, analog_value> analog_reader_values;
+        std::map<std::string, digital_value> digital_interrupt_values;
+    };
```

Comment:
The other sdk's have also chosen the name `status` to indicate this very specific and not-general structure. I am hesitant about this name but have left it in for now. 

I am leaning towards switching to `board_reader_status` but am not sure if that is better. 

The difficult part is that the digital values are `interrupts` but the analog values are `readers` (and GPIO info is not present)

Reply:
It's OK I think. I agree that `status` is a bit of a land grab-y name, but the usage is scoped to `Board` so it isn't that bad.

Reply:
I had an idea:

We could drop the `status` object entirely and delete the `get_status` method. 

Then we could add a `get_analog_readers() -> std::unordered_map<string, analog_value>`, `get_digital_readers() -> ...` which internally calls the `Status()` grpc call and then breaks out the struct.

This would clean up one of the more confusing parts of the API.

Reply:
That would mean though that if you really did want both you would need to make two grpc calls, right? I'm not sure that is a good idea. What if you really are interested in snapshotting both an analog and digital reading at the "same time"?

Reply:
Good point I hadn't considered that.

Reply:
Per discussion, if you want an API that returns just the analog / just the digitals, make some helper functions that call the big function and drop the uninteresting result.

Reply:
@stuqdog Are you also onboard with this?

Reply:
If I do this, I think it makes sense to delete `get_analog_reader_names` and `get_digital_interrrupt_names` and replace then with `get_analog_readers()` and `get_digital_interrupts()` (the former two already call `Status` and then create a vector from the resulting keys of the maps)

Reply:
I think personally I'd keep the `names` ones for alignment with other APIs. The implementation is trivial, just vectorize the map keys.

Reply:
Done in https://github.com/viamrobotics/viam-cpp-sdk/pull/96/commits/c4a1e93b1259b406bc3547d85cce10b79e492fb1. Feel free to unresolve if you have more to discuss.